### ISSUE TITLE
question for embedding your decoder

### ISSUE THREAD

**Issue by publicocean0** [jscodec/jsvpx#12]
hi , i d like embed your decoder in my player. 

but i have some questions
 ipothesis: 
-i have a webm stream ... using chunks.
-the demuxer is seekeable .... it permits to seek in a specific position and send out the right frames.

the decoder is stateless? when i seek there no state to reset in the decoder? nothing related to keyframe for example?

**Comment by brianxautumn** [comment]
The decoder does keep the state. You can only seek to a keyframe, but this is how all decoders work.

**Comment by publicocean0** [comment]
ok pratically after a seek i have to call decode api with compressed frames starting from a keyframe

**Comment by brianxautumn** [comment]
Yes, this is the case with VP8, you have to seek to the nearest keyframe.

**Comment by publicocean0** [comment]
then for the player part ... i think i might refresh canvas with setTimeout every X ms based on sampleRate defined in webm , i m not sure if settimeout is the best way (faster)

**Comment by brianxautumn** [comment]
When you multiplex it, it should put timestamps on each frame. I don't think a solid fps will work perfectly. Do you have audio too?

**Comment by brianxautumn** [comment]
I'm not an expert on A/V sync but perhaps @brion knows

**Comment by brion** [comment]
Indeed there's no single fps value provided in WebM structure; each frame's packet lists a presentation timestamp relative to the start time.

Very roughly a playback loop (such as is implemented in [my ogv.js player](https://github.com/brion/ogv.js) needs to do:

* demux next frame
* get the current playback time, something like:
    * `playbackTime = audioContext.currentTime` if you have audio, or just:
    * `playbackTime = Date.now() - startTime` or if you're really fancy and want to handle pauses better:
    * `playbackTime = (Date.now() - timeWeLastPlayedAFrame) + ptsOfLastFrame`
* delay until it's time
    * `setTimeout(displayFrame, ptsOfCurrentFrame - playbackTime)`
* draw!
* go back to start

**Comment by publicocean0** [comment]
Thanks a lot @brion . Answering to @brianxautumn. No i have still to see exactly how to do . I have just saw you realized a audiostream feeder for doing it. 

The better solution would be realize a MSE object for browser not supporting it. 
There is also another little problem to think. 
Usually there is a parser/demuxer also in dash component. 
MSE might have also a simplifed version of a demuxer+ decoder. 
There is a rendundancy . I m asking if is better to think a trick for evoiding it , but anyway realizing a portable solution.
